### PR TITLE
[Arctic-1311] Add new task splitting rules

### PR DESCRIPTION
**Pull request by shidayang**
## Why are the changes needed?
Resolve: #1311 

## Brief change log
Add new task splitting rules in BasicKeyedTableScan

In benchmark, When all the data is static, the query time is reduced from 19.9s to 14s. 

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: core/src/main/java/com/netease/arctic/scan/BasicKeyedTableScan.java [hunk 1]
```diff
@@ -149,16 +160,40 @@ private CloseableIterable<ArcticFileScanTask> planChangeFiles() {
...
+              .sum();
+
+          long dataWeight = task.dataTasks().stream().mapToLong(s -> s.file().fileSizeInBytes())
+              .map(s -> s + openFileCost)
+              .sum();
+          double deleteRatio = deleteWeight * 1.0 / dataWeight;
+
+          if (deleteRatio < splitTaskByDeleteRatio) {
+            long targetSize = Math.min(new Double(deleteWeight / splitTaskByDeleteRatio).longValue(), splitSize);
+            CloseableIterable<NodeFileScanTask> tasksIterable =
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/scan/BasicKeyedTableScan.java:180]: Does the `enableSplitTaskByDelete` only affect the generation of `targetSize`, and is there any consideration for reusing the subsequent split code?

File: core/src/main/java/com/netease/arctic/utils/ManifestEntryFields.java [hunk 2]
```diff
@@ -27,6 +27,9 @@ public class ManifestEntryFields {
   public static final Types.NestedField STATUS = required(0, "status", Types.IntegerType.get());
   public static final Types.NestedField SNAPSHOT_ID = optional(1, "snapshot_id", Types.LongType.get());
   public static final Types.NestedField SEQUENCE_NUMBER = optional(3, "sequence_number", Types.LongType.get());
+
+  public static  final Types.NestedField FILE_SEQUENCE_NUMBER =
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/utils/ManifestEntryFields.java:31]: Does this PR request this field?

File: site/docs/ch/configurations.md [hunk 3]
```diff
@@ -44,11 +44,12 @@ Self-optimizing 配置对 Iceberg format, Mixed streaming format 都会生效。
...
+| ---------------------------------- | ---------------- |----------------------------------------------------------------------------------------------------------------------------------------------------------|
+| read.split.open-file-cost          | 4194304（4MB）    | 预估与读取开销等价的打开一个文件开销，一般不改                                                                                                                                  |
+| read.split.planning-lookback       | 10               | 拆分读取任务所使用分桶算法里桶的个数                                                                                                                                       |
+| read.split.target-size              | 134217728（128MB）| 查询引擎中的读取任务会尽量拆分成这个大小                                                                                                                                     |
+| read.split.delete-ratio            | 0.05             | 在 Mix-format 中当 Node 中的 delete 数据占比小于此值，那么就会将 Node 切分为多个 Task，且切分后的 Task 的 delete 数据占比不得大于此值，不过切分过后的 Task 的 data size 不会超过 read.split.target-size 所设置的大小 |
```
Comment (zhoujinsong) [inline site/docs/ch/configurations.md:52]: How about change the description to:
```
Delete文件占比小于此阈值时，读取任务会拆分成更多的任务以提高查询速度。
```

File: core/src/main/java/com/netease/arctic/scan/KeyedTableScan.java [hunk 4]
```diff
@@ -41,4 +41,9 @@ public interface KeyedTableScan {
    */
   CloseableIterable<CombinedScanTask> planTasks();
 
+  /**
+   * Config if to split task by delete ratio, default is false
+   */
+  KeyedTableScan enableSplitTaskByDeleteRatio();
```
[the remark this answers is not in the document]
Comment (shidayang) [reply core/src/main/java/com/netease/arctic/scan/KeyedTableScan.java:47]: Current only for Trino use.

File: core/src/main/java/com/netease/arctic/utils/ManifestEntryFields.java (hunk 2 above)
Comment (shidayang) [reply core/src/main/java/com/netease/arctic/utils/ManifestEntryFields.java:31]: Really don't need this

File: site/docs/ch/configurations.md (hunk 3 above)
Comment (shidayang) [reply site/docs/ch/configurations.md:52]: agree

**Comment by shidayang** [comment]
@zhoujinsong I made some modifications

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 5]
```diff
@@ -247,6 +247,10 @@ private TableProperties() {
   public static final String SPLIT_SIZE = org.apache.iceberg.TableProperties.SPLIT_SIZE;
   public static final long SPLIT_SIZE_DEFAULT = 134217728; // 128 MB
 
+  public static final String SPLIT_TASK_BY_DELETE_RATIO = "read.split.delete-ratio";
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/table/TableProperties.java:250]: Currently, this feature is only effective within Trino. Is it appropriate to directly place the parameters on the table?

**Review by zhoujinsong — APPROVED** [review]
LGTM.

File: core/src/main/java/com/netease/arctic/scan/BasicKeyedTableScan.java [hunk 6]
```diff
@@ -121,13 +122,20 @@ public CloseableIterable<CombinedScanTask> planTasks() {
         splitSize, lookBack, openFileCost);
   }
 
+  @Override
+  public KeyedTableScan enableSplitTaskByDeleteRatio(double splitTaskByDeleteRatio) {
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/scan/BasicKeyedTableScan.java:126]: May need to add some unit test for this new feature.

File: core/src/main/java/com/netease/arctic/scan/KeyedTableScan.java [hunk 7]
```diff
@@ -41,4 +41,9 @@ public interface KeyedTableScan {
    */
   CloseableIterable<CombinedScanTask> planTasks();
 
+  /**
+   * Config if to split task by delete ratio, default is false
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/scan/KeyedTableScan.java:45]: Is this javadoc out of date?

File: core/src/test/java/com/netease/arctic/scan/TestScanSplitTask.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+          getArcticTable().asKeyedTable().newScan().planTasks();
+      for (CombinedScanTask combinedScanTask : combinedScanTasks) {
+        Assert.assertEquals(combinedScanTask.tasks().size(), 7);
+      }
+    }
+    {
+      CloseableIterable<CombinedScanTask> combinedScanTasks =
+          getArcticTable().asKeyedTable().newScan().enableSplitTaskByDeleteRatio(0.04).planTasks();
+      for (CombinedScanTask combinedScanTask : combinedScanTasks) {
+        Assert.assertEquals(combinedScanTask.tasks().size(), 9);
```
Comment (zhoujinsong) [inline core/src/test/java/com/netease/arctic/scan/TestScanSplitTask.java:52]: Add some comments to explain why the scan should be split into 9 tasks here.

**Pull request merged**